pkg/profiler/profiler.go

```diff
@@ -229,80 +249,74 @@ func (p *CgroupProfiler) Run(ctx context.Context) error {
+		prof = &profile.Profile{
+			SampleType: []*profile.ValueType{{
+				Type: "samples",
+				Unit: "count",
+			}},
+			TimeNanos:     now.UnixNano(),
+			DurationNanos: int64(p.profilingDuration),
+
+			// We sample at 100Hz, which is every 10 Million nanoseconds.
+			PeriodType: &profile.ValueType{
+				Type: "cpu",
+				Unit: "nanoseconds",
+			},
+			Period: 10000000,
+		}
 
-	// TODO(brancz): What was this for?
-	//has_collision := false
+		mapping       = maps.NewMapping(p.pidMappingFileCache)
+		kernelMapping = &profile.Mapping{
+			File: "[kernel.kallsyms]",
+		}
+		kernelFunctions = map[uint64]*profile.Function{}
+		userFunctions   = map[[2]uint64]*profile.Function{}
 
-	it := counts.Iterator()
-	byteOrder := byteorder.GetHostByteOrder()
+		// 2 uint64 1 for PID and 1 for Addr
+		locations       = []*profile.Location{}
+		kernelLocations = []*profile.Location{}
+		kernelAddresses = map[uint64]struct{}{}
+		locationIndices = map[[2]uint64]int{}
+		samples         = map[[doubleStackDepth]uint64]*profile.Sample{}
+		byteOrder       = byteorder.GetHostByteOrder()
 
-	// TODO(brancz): Use libbpf batch functions.
-	for it.Next() {
-		// This byte slice is only valid for this iteration, so it must be
-		// copied if we want to do anything with it outside of this loop.
-		keyBytes := it.Key()
+		// Variables needed for eBPF map batch iteration.
+		keysPtr = unsafe.Pointer(&keys[0])
+		nextKey = uintptr(1)
+	)
 
-		r := bytes.NewBuffer(keyBytes)
+	memsetCountKeys(keys, stackCountKey{})
 
-		pidBytes := make([]byte, 4)
-		if _, err := io.ReadFull(r, pidBytes); err != nil {
-			return fmt.Errorf("read pid bytes: %w", err)
+	vals, err := counts.GetValueAndDeleteBatch(keysPtr, nil, unsafe.Pointer(&nextKey), counts.GetMaxEntries())
+	if err != nil {
+		if !errors.Is(err, syscall.ENOENT) { // Map is empty or we got all keys in the last batch.
+			return err
 		}
-		pid := byteOrder.Uint32(pidBytes)
+	}
 
-		userStackIDBytes := make([]byte, 4)
-		if _, err := io.ReadFull(r, userStackIDBytes); err != nil {
-			return fmt.Errorf("read user stack ID bytes: %w", err)
-		}
-		userStackID := int32(byteOrder.Uint32(userStackIDBytes))
+	for i, key := range keys {
+		var (
+			pid           = key.pid
+			userStackID   = key.userStackID
+			kernelStackID = key.kernelStackID
+		)
 
-		kernelStackIDBytes := make([]byte, 4)
-		if _, err := io.ReadFull(r, kernelStackIDBytes); err != nil {
-			return fmt.Errorf("read kernel stack ID bytes: %w", err)
+		if pid == 0 {
+			break
 		}
-		kernelStackID := int32(byteOrder.Uint32(kernelStackIDBytes))
 
-		valueBytes, err := counts.GetValue(unsafe.Pointer(&keyBytes[0]))
-		if err != nil {
-			return fmt.Errorf("get count value: %w", err)
-		}
-		value := byteOrder.Uint64(valueBytes)
+		value := byteOrder.Uint64(vals[i])
 
 		stackBytes, err := stackTraces.GetValue(unsafe.Pointer(&userStackID))
 		if err != nil {
-			//profile.MissingStacks++
+			// TODO(derekparker): Should we log or increment missing stack trace count?
```

Comment:
:+1: Let's add a metric for it, not necessarily in this PR though. https://github.com/parca-dev/parca-agent/issues/208